File: notify.php
<?php

error_reporting(E_ALL);
ini_set('display_errors', 1);

// ===== 設定 =====
$webhook     = trim(getenv('DISCORD_WEBHOOK') ?: '');
$steamApiKey = trim(getenv('STEAM_API_KEY') ?: '');
$minReviews  = 50;
$seedPages   = 10;
$dataFile    = __DIR__ . '/languages.json';
$metaFile    = __DIR__ . '/meta.json';
$chineseKeys = ['Simplified Chinese', 'Traditional Chinese'];

if (!$webhook)     exit('[ERROR] DISCORD_WEBHOOK 未設定' . PHP_EOL);
if (!$steamApiKey) exit('[ERROR] STEAM_API_KEY 未設定' . PHP_EOL);

// ===== HTTP 工具函式（curl，含逾時與狀態碼檢查）=====
function httpGet(string $url, int $timeout = 15): ?string {
    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT        => $timeout,
        CURLOPT_FOLLOWLOCATION => true,
        CURLOPT_USERAGENT      => 'Mozilla/5.0 (compatible; SteamChineseNotifier/1.0)',
    ]);
    $result = curl_exec($ch);
    $code   = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error  = curl_error($ch);
    curl_close($ch);

    if ($error) {
        echo "[WARN] curl 錯誤：{$error}" . PHP_EOL;
        return null;
    }
    if ($code !== 200) {
        echo "[WARN] HTTP {$code}：{$url}" . PHP_EOL;
        return null;
    }
    return $result ?: null;
}

// ===== 讀取 meta & 語系資料 =====
$meta   = file_exists($metaFile) ? (json_decode(file_get_contents($metaFile), true) ?? []) : [];
$stored = file_exists($dataFile) ? (json_decode(file_get_contents($dataFile), true) ?? []) : [];

// 確保檔案一定存在（讓 git add 不會報錯）
if (!file_exists($dataFile)) file_put_contents($dataFile, '{}');
if (!file_exists($metaFile)) file_put_contents($metaFile, '{}');

$lastChangeNumber = (int)($meta['last_change_number'] ?? 0);
$isFirstRun       = ($lastChangeNumber === 0);

// ===== 取得 Steam change number =====
echo '[INFO] 查詢 Steam GetAppChanges...' . PHP_EOL;
$changesRaw = httpGet(
    "https://api.steampowered.com/ISteamApps/GetAppChanges/v2/?changeNumber={$lastChangeNumber}&key={$steamApiKey}"
);

if (!$changesRaw) {
    exit('[ERROR] 無法連接 Steam GetAppChanges API，請確認 STEAM_API_KEY 是否正確' . PHP_EOL);
}

$changesData = json_decode($changesRaw, true);
if ($changesData === null) {
    exit('[ERROR] Steam API 回傳無效 JSON：' . $changesRaw . PHP_EOL);
}

$currentChangeNumber = (int)($changesData['current_change_number'] ?? $lastChangeNumber);
$forceFullUpdate     = !empty($changesData['force_full_update']);

// GetAppChanges 可能回傳 object 或 array，統一處理
$rawChangedApps = $changesData['appchanges'] ?? [];
$changedApps    = [];
if (is_array($rawChangedApps)) {
    foreach ($rawChangedApps as $key => $val) {
        // object 格式：{"570": {...}}  或  array 格式：[{"appid":570,...}]
        $appid = is_numeric($key) && isset($val['appid']) ? (string)$val['appid'] : (string)$key;
        $changedApps[$appid] = $val;
    }
}

echo "[INFO] change number：{$lastChangeNumber} → {$currentChangeNumber}" . PHP_EOL;
echo '[INFO] 有變動 App：' . count($changedApps) . ' 個' . PHP_EOL;

// 立即儲存最新 change number
$meta['last_change_number'] = $currentChangeNumber;
$meta['last_run']           = date('Y-m-d H:i:s');
file_put_contents($metaFile, json_encode($meta, JSON_PRETTY_PRINT));

// =========================================================
// 首次執行：用 SteamSpy 掃描前 10,000 款遊戲建立基準
// =========================================================
if ($isFirstRun || $forceFullUpdate) {
    echo '[INFO] 首次執行，掃描 SteamSpy 前 ' . ($seedPages * 1000) . ' 款遊戲...' . PHP_EOL;
    echo '[INFO] 本次不推播，僅建立基準資料' . PHP_EOL;

    $allGames = [];
    for ($page = 0; $page < $seedPages; $page++) {
        echo '[INFO] SteamSpy 第 ' . ($page + 1) . "/{$seedPages} 頁..." . PHP_EOL;

        $raw = httpGet("https://steamspy.com/api.php?request=all&page={$page}", 20);
        if (!$raw) {
            echo "[WARN] 第 {$page} 頁取得失敗，停止翻頁" . PHP_EOL;
            break;
        }

        $games = json_decode($raw, true);
        if (!is_array($games) || empty($games)) {
            echo '[INFO] 已無更多資料' . PHP_EOL;
            break;
        }

        $allGames = array_merge($allGames, $games);
        sleep(2);
    }

    $filtered = array_filter($allGames, function ($g) use ($minReviews) {
        return isset($g['positive'], $g['negative'])
            && ($g['positive'] + $g['negative']) > $minReviews;
    });

    echo '[INFO] 評論數 > ' . $minReviews . ' 的遊戲：' . count($filtered) . ' 款' . PHP_EOL;
    echo '[INFO] 預計需要約 ' . round(count($filtered) * 0.5 / 60) . ' 分鐘...' . PHP_EOL;

    $i = 0; $total = count($filtered);
    foreach ($filtered as $game) {
        if (!isset($game['appid'])) continue;
        $appid = (string)$game['appid'];
        $i++;

        if (isset($stored[$appid])) continue;

        echo "[{$i}/{$total}] {$game['name']} ({$appid})" . PHP_EOL;

        $raw = httpGet("https://store.steampowered.com/api/appdetails?appids={$appid}&filters=supported_languages");
        if (!$raw) { usleep(500000); continue; }

        $data = json_decode($raw, true);
        if (!is_array($data) || empty($data[$appid]['success'])) { usleep(500000); continue; }

        $languages  = $data[$appid]['data']['supported_languages'] ?? '';
        $hasChinese = false;
        foreach ($chineseKeys as $key) {
            if (stripos($languages, $key) !== false) { $hasChinese = true; break; }
        }

        $stored[$appid] = [
            'name'         => $game['name'] ?? "App {$appid}",
            'has_chinese'  => $hasChinese,
            'last_checked' => date('Y-m-d'),
        ];

        echo '  → 中文：' . ($hasChinese ? '✅ 有' : '❌ 無') . PHP_EOL;

        if ($i % 100 === 0) {
            file_put_contents($dataFile, json_encode($stored, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
            echo "[INFO] 進度儲存（{$i}/{$total}）" . PHP_EOL;
        }

        usleep(500000);
    }

    file_put_contents($dataFile, json_encode($stored, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
    echo '[INFO] 基準建立完成，共記錄 ' . count($stored) . ' 款遊戲 ✅' . PHP_EOL;
    exit(0);
}

// =========================================================
// 後續執行：追蹤有更新的遊戲
// =========================================================
if (empty($changedApps)) {
    echo '[INFO] 本次無任何 App 更新' . PHP_EOL;
    exit(0);
}

$newChineseList = [];
$notified       = [];   // 防止同一遊戲重複推播
$i = 0; $total = count($changedApps);

foreach ($changedApps as $appid => $changeData) {
    $appid = (string)$appid;
    $i++;

    if (isset($stored[$appid]) && ($stored[$appid]['has_chinese'] ?? false)) continue;

    echo "[{$i}/{$total}] 查詢 App {$appid}..." . PHP_EOL;

    $raw = httpGet("https://store.steampowered.com/api/appdetails?appids={$appid}");
    if (!$raw) { usleep(500000); continue; }

    $data = json_decode($raw, true);
    if (!is_array($data) || empty($data[$appid]['success'])) { usleep(500000); continue; }

    $appData = $data[$appid]['data'] ?? [];
    if (empty($appData)) { usleep(300000); continue; }

    $type = $appData['type'] ?? '';
    $name = $appData['name'] ?? "App {$appid}";

    if ($type !== 'game') {
        echo "  → [{$name}] 非遊戲（{$type}），跳過" . PHP_EOL;
        usleep(300000);
        continue;
    }

    $totalReviews = $appData['recommendations']['total'] ?? 0;
    if ($totalReviews < $minReviews) {
        echo "  → [{$name}] 評論數 {$totalReviews} < {$minReviews}，跳過" . PHP_EOL;
        usleep(300000);
        continue;
    }

    $languages  = $appData['supported_languages'] ?? '';
    $hasChinese = false;
    foreach ($chineseKeys as $key) {
        if (stripos($languages, $key) !== false) { $hasChinese = true; break; }
    }

    $wasTracked = isset($stored[$appid]);
    $hadChinese = $wasTracked && ($stored[$appid]['has_chinese'] ?? false);

    if ($hasChinese && $wasTracked && !$hadChinese && !isset($notified[$appid])) {
        echo "  → 🎉 [{$name}] 新增中文支援！" . PHP_EOL;

        $reviewRaw  = httpGet("https://store.steampowered.com/appreviews/{$appid}?json=1&language=all&purchase_type=all");
        $reviewJson = $reviewRaw ? json_decode($reviewRaw, true) : null;
        $reviewData = is_array($reviewJson) ? ($reviewJson['query_summary'] ?? []) : [];

        $newChineseList[] = [
            'appid'    => $appid,
            'name'     => $name,
            'languages'=> strip_tags($languages),
            'positive' => (int)($reviewData['total_positive'] ?? 0),
            'negative' => (int)($reviewData['total_negative'] ?? 0),
            'total'    => $totalReviews,
        ];
        $notified[$appid] = true;
    } else {
        echo "  → [{$name}] 中文：" . ($hasChinese ? '✅ 有' : '❌ 無') . "（評論：{$totalReviews}）" . PHP_EOL;
    }

    $stored[$appid] = [
        'name'         => $name,
        'has_chinese'  => $hasChinese,
        'last_checked' => date('Y-m-d'),
    ];

    usleep(500000);
}

file_put_contents($dataFile, json_encode($stored, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
echo '[INFO] languages.json 已更新，共記錄 ' . count($stored) . ' 款遊戲' . PHP_EOL;

// ===== 推播 Discord =====
if (empty($newChineseList)) {
    echo '[INFO] 本次無新增中文的遊戲' . PHP_EOL;
    exit(0);
}

echo '[INFO] 推播 ' . count($newChineseList) . ' 款遊戲到 Discord...' . PHP_EOL;

foreach ($newChineseList as $game) {
    $reviewTotal = $game['positive'] + $game['negative'];
    $rate        = $reviewTotal > 0 ? round($game['positive'] / $reviewTotal * 100) : 0;
    $langText    = mb_substr($game['languages'], 0, 1000);  // Discord field 上限 1024

    $payload = [
        'embeds' => [[
            'title'     => mb_substr("🎮 新增中文支援：{$game['name']}", 0, 256),
            'url'       => "https://store.steampowered.com/app/{$game['appid']}/",
            'color'     => 5763719,
            'fields'    => [
                ['name' => '📊 評論', 'value' => "👍 {$game['positive']}  👎 {$game['negative']}（好評率 {$rate}%）", 'inline' => false],
                ['name' => '🌐 支援語系', 'value' => $langText ?: '（無資料）', 'inline' => false],
            ],
            'timestamp' => date('c'),
        ]],
    ];

    $ch = curl_init($webhook);
    curl_setopt_array($ch, [
        CURLOPT_POST           => true,
        CURLOPT_POSTFIELDS     => json_encode($payload, JSON_UNESCAPED_UNICODE),
        CURLOPT_HTTPHEADER     => ['Content-Type: application/json'],
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT        => 10,
    ]);
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    if ($httpCode === 429) {
        echo "[WARN] Discord 速率限制，等待 5 秒..." . PHP_EOL;
        sleep(5);
    } elseif ($httpCode >= 400) {
        echo "[WARN] Discord 推播失敗（HTTP {$httpCode}）：{$game['name']}" . PHP_EOL;
    } else {
        echo "[NOTIFY] 已推播：{$game['name']}" . PHP_EOL;
    }

    sleep(2);
}

File: notify.py
"""
notify.py — 每日偵測新增繁/簡體中文支援的 Steam 遊戲並推播到 Discord

流程：
  1. IStoreService/GetAppList 取得自上次執行以來有變動的遊戲清單
  2. 逐一查詢 Store API，比對語系是否從「無中文 → 有中文」
  3. 符合條件者推播 Discord embed 通知
"""
import json
import os
import re
import time
from datetime import datetime, timezone, timedelta
from pathlib import Path

import requests

# ===== 設定 =====
WEBHOOK       = os.environ.get('DISCORD_WEBHOOK', '').strip()
STEAM_API_KEY = os.environ.get('STEAM_API_KEY', '').strip()
MIN_REVIEWS   = 20
DATA_FILE     = Path(__file__).parent / 'languages.json'
META_FILE     = Path(__file__).parent / 'meta.json'

CHINESE_STORE_KEYS = {'Simplified Chinese', 'Traditional Chinese', '簡體中文', '繁體中文'}

if not WEBHOOK:
    raise SystemExit('[ERROR] DISCORD_WEBHOOK 未設定')
if not STEAM_API_KEY:
    raise SystemExit('[ERROR] STEAM_API_KEY 未設定')
if not DATA_FILE.exists():
    raise SystemExit('[ERROR] languages.json 不存在，請先執行 seed.py 建立基準')


# ===== 讀取狀態檔 =====
meta   = json.loads(META_FILE.read_text('utf-8')) if META_FILE.exists() else {}
stored = json.loads(DATA_FILE.read_text('utf-8'))

# 上次執行時間，預設為 25 小時前確保不漏掉
last_run_ts = int(meta.get('last_run_ts', 0))
if not last_run_ts:
    last_run_ts = int((datetime.now(timezone.utc) - timedelta(hours=25)).timestamp())

print(f'[INFO] 查詢 {datetime.fromtimestamp(last_run_ts, timezone.utc).strftime("%Y-%m-%d %H:%M")} 之後有變動的遊戲...', flush=True)


# ===== 透過 IStoreService/GetAppList 取得有變動的遊戲 =====
changed_appids = []
last_appid = 0

while True:
    params = {
        'key':               STEAM_API_KEY,
        'if_modified_since': last_run_ts,
        'include_games':     'true',
        'max_results':       50000,
    }
    if last_appid:
        params['last_appid'] = last_appid

    try:
        r = requests.get('https://api.steampowered.com/IStoreService/GetAppList/v1/', params=params, timeout=30)
        r.raise_for_status()
        data = r.json().get('response', {})
        apps = data.get('apps', [])
        changed_appids.extend(a['appid'] for a in apps)

        if not data.get('have_more_results'):
            break
        last_appid = data.get('last_appid', 0)
        time.sleep(1)
    except Exception as e:
        print(f'[WARN] GetAppList 失敗：{e}')
        break

print(f'[INFO] 有變動 App：{len(changed_appids)} 個')

# 更新執行時間戳記（在查詢前更新，避免重複通知）
now_ts = int(datetime.now(timezone.utc).timestamp())
meta['last_run_ts'] = now_ts
meta['last_run']    = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
META_FILE.write_text(json.dumps(meta, indent=2), 'utf-8')

if not changed_appids:
    print('[INFO] 本次無任何 App 更新')
    raise SystemExit(0)

SKIP_TYPES = {'dlc', 'music', 'demo', 'advertising', 'mod', 'video', 'free'}

to_check = [appid for appid in changed_appids
            if not stored.get(str(appid), {}).get('has_chinese')
            and stored.get(str(appid), {}).get('app_type', 'game') not in SKIP_TYPES]

print(f'[INFO] 實際需查詢：{len(to_check)} 個（跳過已知有中文）')

STORE_SLEEP = 2.0  # 每筆間隔秒數，避免觸發 Store API 速率限制


# ===== 逐一用 Store API 查詢語系 =====
def get_store_info(appid: int) -> tuple[str | None, str | None, list, dict]:
    """回傳 (app_type, name, langs, extra)；查詢失敗時 app_type 為 None。"""
    url  = f'https://store.steampowered.com/api/appdetails?appids={appid}&l=tchinese'
    wait = 60
    for _ in range(3):
        try:
            r = requests.get(url, timeout=15)
            if r.status_code == 429:
                print(f'  [WARN] 速率限制，等待 {wait} 秒...', flush=True)
                time.sleep(wait)
                wait *= 2
                continue
            data = r.json()
            if not data or not data.get(str(appid), {}).get('success'):
                return None, None, [], {}
            app_data = data[str(appid)]['data']
            release  = app_data.get('release_date', {})
            if release.get('coming_soon') or not release.get('date'):
                return 'coming_soon', app_data.get('name', f'App {appid}'), [], {}
            app_type  = app_data.get('type', '').lower()
            name      = app_data.get('name', f'App {appid}')
            # split('*')[0] 去掉星號（部分語音支援標記）及後方的備註文字
            raw_langs = app_data.get('supported_languages', '')
            langs     = [l.strip().split('*')[0].strip() for l in re.sub(r'<[^>]+>', '', raw_langs).split(',')]
            extra = {
                'genres':       ', '.join(g['description'] for g in app_data.get('genres', [])),
                'developers':   ', '.join(app_data.get('developers', [])),
                'release_date': release.get('date', ''),
                'is_free':      app_data.get('is_free', False),
                'header_image': app_data.get('header_image', ''),
            }
            return app_type, name, langs, extra
        except Exception as e:
            print(f'  [WARN] Store API 失敗：{e}', flush=True)
            return None, None, [], {}
    return None, None, [], {}


new_chinese_list = []
notified = set()
today    = datetime.now().strftime('%Y-%m-%d')

for i, appid in enumerate(to_check, 1):
    appid_str = str(appid)
    print(f'[{i}/{len(to_check)}] 查詢 App {appid}...', flush=True)

    app_type, name, langs, extra = get_store_info(appid)

    if app_type is None:
        time.sleep(STORE_SLEEP)
        continue

    if app_type != 'game':
        label = '尚未發售' if app_type == 'coming_soon' else f'非遊戲（{app_type}）'
        print(f'  → {label}，跳過')
        stored[appid_str] = {
            'name': name, 'has_chinese': False, 'app_type': app_type, 'last_checked': today,
        }
        time.sleep(STORE_SLEEP)
        continue

    if extra.get('is_free'):
        print(f'  → 免費遊戲，跳過')
        stored[appid_str] = {'name': name, 'has_chinese': False, 'app_type': 'free', 'last_checked': today}
        time.sleep(STORE_SLEEP)
        continue

    has_chinese = bool(set(langs) & CHINESE_STORE_KEYS)
    had_chinese = stored.get(appid_str, {}).get('has_chinese', False)

    if has_chinese and not had_chinese and appid_str not in notified:
        print(f'  → 🎉 [{name}] 新增中文支援！')

        total_reviews = positive = negative = 0
        try:
            r = requests.get(
                f'https://store.steampowered.com/appreviews/{appid}?json=1&language=all&purchase_type=all',
                timeout=10,
            )
            summary       = r.json().get('query_summary', {})
            positive      = summary.get('total_positive', 0)
            negative      = summary.get('total_negative', 0)
            total_reviews = summary.get('total_reviews', positive + negative)
        except Exception:
            pass

        if total_reviews < MIN_REVIEWS:
            print(f'  → 評論數 {total_reviews} < {MIN_REVIEWS}，跳過')
        else:
            new_chinese_list.append({
                'appid':        appid_str,
                'name':         name,
                'langs':        ', '.join(l for l in langs if l),
                'positive':     positive,
                'negative':     negative,
                'genres':       extra.get('genres', ''),
                'developers':   extra.get('developers', ''),
                'release_date': extra.get('release_date', ''),
                'header_image': extra.get('header_image', ''),
            })
            notified.add(appid_str)
    else:
        print(f'  → 中文：{"✅ 有" if has_chinese else "❌ 無"}')

    stored[appid_str] = {
        'name': name, 'has_chinese': has_chinese, 'last_checked': today,
    }
    time.sleep(STORE_SLEEP)


DATA_FILE.write_text(json.dumps(stored, indent=2, ensure_ascii=False), 'utf-8')
print(f'[INFO] languages.json 已更新，共記錄 {len(stored)} 款遊戲')


# ===== 推播 Discord =====
if not new_chinese_list:
    print('[INFO] 本次無新增中文的遊戲')
    raise SystemExit(0)

print(f'[INFO] 推播 {len(new_chinese_list)} 款遊戲到 Discord...')

BATCH_SIZE  = 10
total       = len(new_chinese_list)
today_str   = datetime.now().strftime('%Y-%m-%d')
batches     = [new_chinese_list[i:i + BATCH_SIZE] for i in range(0, total, BATCH_SIZE)]
batch_count = len(batches)

for idx, batch in enumerate(batches):
    embeds = []
    for game_idx, game in enumerate(batch):
        review_total = game['positive'] + game['negative']
        rate  = round(game['positive'] / review_total * 100) if review_total > 0 else 0
        url   = f"https://store.steampowered.com/app/{game['appid']}/"
        img   = game.get('header_image') or f"https://cdn.akamai.steamstatic.com/steam/apps/{game['appid']}/header.jpg"
        part_str = f'（第 {idx + 1} / {batch_count} 則）' if batch_count > 1 else ''

        is_first = game_idx == 0
        title    = game['name'][:256]
        if idx == 0 and is_first:
            author_name = f'🌏 今日新增中文支援：共 {total} 款遊戲 {part_str}'.strip()
        elif is_first:
            author_name = f'🌏 新增中文支援（續）{part_str}'.strip()
        else:
            author_name = None

        fields = [
            {'name': '📊 評論', 'value': f"👍 {game['positive']}  👎 {game['negative']}（好評率 {rate}%）", 'inline': False},
        ]
        if game.get('genres'):
            fields.append({'name': '🏷️ 類型',   'value': game['genres'],       'inline': True})
        if game.get('developers'):
            fields.append({'name': '🛠️ 開發商', 'value': game['developers'],   'inline': True})
        if game.get('release_date'):
            fields.append({'name': '📅 發售日', 'value': game['release_date'], 'inline': True})

        embed = {
            'title':     title[:256],
            'url':       url,
            'color':     5763719,
            'image':     {'url': img},
            'fields':    fields,
            'footer':    {'text': f'Steam 中文更新通知 · {today_str}'},
            'timestamp': datetime.now(timezone.utc).isoformat(),
        }
        if author_name:
            embed['author'] = {'name': author_name}
        embeds.append(embed)

    payload = {'embeds': embeds}
    r = requests.post(WEBHOOK, json=payload, timeout=10)

    if r.status_code == 429:
        print(f'[WARN] Discord 速率限制，等待 5 秒...')
        time.sleep(5)
        r = requests.post(WEBHOOK, json=payload, timeout=10)

    if r.status_code >= 400:
        print(f'[WARN] Discord 推播失敗（HTTP {r.status_code}，第 {idx + 1} 則）')
    else:
        print(f'[NOTIFY] 已推播第 {idx + 1} 則（{len(batch)} 款）')

    if idx < batch_count - 1:
        time.sleep(2)
